File: 리트코드/0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        List<List<Integer>> result = new ArrayList<>();
        List<Integer> collect = Arrays.stream(candidates).boxed().collect(Collectors.toList());
        dfs(result, collect, new LinkedList<>(),0, target, collect.size(), 0, 0);

        return result;
    }

    private void dfs(List<List<Integer>> result, List<Integer> collect,
            LinkedList<Integer> elements,
            int start, int target, int n, int k, int sum) {

        if (sum > target) {
            return;
        }
        if (sum == target) {
            result.add(elements.stream().collect(Collectors.toList()));
            return;
        }
        for (int i = start; i < n; i++) {
            int val = collect.get(i);
            elements.add(val);
            dfs(result, collect, elements,i, target, n, k + 1, sum + val);
            elements.removeLast();
        }
    }
}

File: 리트코드/0039-combination-sum/0039-combination-sum.kt
class Solution {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val result: MutableList<List<Int>> = mutableListOf()
        fun dfs(path: ArrayDeque<Int>, index: Int, target: Int) {

            if (target < 0) {
                return
            }
            if (target == 0) {
                result.add(path.stream().collect(Collectors.toList()))
                return
            }

            for (i in index until candidates.size) {
                path.add(candidates[i])
                dfs(path, i, target - candidates[i])
                path.removeLast()
            }
        }

        dfs(ArrayDeque(), 0, target)
        return result
    }
}